components/file_system/src/rate_limiter.rs

```diff
@@ -1,43 +1,363 @@
+            }
+            IOOp::Write => {
+                self.write[io_type as usize].fetch_add(len, Ordering::Relaxed);
+            }
+        }
+    }
+
+    pub fn fetch(&self, io_type: IOType, io_op: IOOp) -> usize {
+        match io_op {
+            IOOp::Read => self.read[io_type as usize].load(Ordering::Relaxed),
+            IOOp::Write => self.write[io_type as usize].load(Ordering::Relaxed),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn reset(&self) {
+        for i in self.read.iter() {
+            i.store(0, Ordering::Relaxed);
+        }
+        for i in self.write.iter() {
+            i.store(0, Ordering::Relaxed);
+        }
+    }
+}
+
+#[derive(Debug)]
+struct PerTypeIORateLimiter {
+    bytes_per_refill: AtomicUsize,
+    consumed: AtomicUsize,
+    refill_period: Duration,
+    last_refill_time: Mutex<Instant>,
+    condv: Condvar,
+    calibrator: Option<BytesCalibrator>,
+}
+
+#[inline]
+fn calculate_bytes_per_refill(bytes_per_sec: usize, refill_period: Duration) -> usize {
+    (bytes_per_sec as f64 * refill_period.as_secs_f64()) as usize
+}
+
+impl PerTypeIORateLimiter {
+    /// Create a new rate limiter. Rate limiting is disabled when `bytes_per_sec` is zero.
+    pub fn new(bytes_per_sec: usize, refill_period: Duration) -> PerTypeIORateLimiter {
+        PerTypeIORateLimiter {
+            bytes_per_refill: AtomicUsize::new(calculate_bytes_per_refill(
+                bytes_per_sec,
+                refill_period,
+            )),
+            consumed: AtomicUsize::new(0),
+            refill_period,
+            last_refill_time: Mutex::new(Instant::now()),
+            condv: Condvar::new(),
+            calibrator: None,
+        }
+    }
+
+    pub fn set_bytes_per_sec(&self, bytes_per_sec: usize) {
+        self.bytes_per_refill.store(
+            calculate_bytes_per_refill(bytes_per_sec, self.refill_period),
+            Ordering::Relaxed,
+        );
+    }
+
+    #[allow(dead_code)]
+    pub fn set_calibrator(&mut self, calibrator: BytesCalibrator) {
+        self.calibrator = Some(calibrator);
+    }
+
+    #[inline]
+    fn request_fast(&self, bytes_per_refill: usize, bytes: usize) -> Option<usize> {
+        if self.consumed.load(Ordering::Relaxed) < bytes_per_refill {
+            // Consumed bytes are allowed to be larger than the actual bytes
+            // through when quotas are drained.
+            let before = self.consumed.fetch_add(bytes, Ordering::Relaxed);
+            if before < bytes_per_refill {
+                return Some(std::cmp::min(bytes_per_refill - before, bytes));
+            }
+        }
+        None
+    }
+
+    #[inline]
+    fn refill_and_request(&self, bytes: usize) -> usize {
+        let token = std::cmp::min(self.bytes_per_refill.load(Ordering::Relaxed), bytes);
+        self.consumed.store(token, Ordering::Relaxed);
+        if let Some(calibrator) = &self.calibrator {
+            calibrator.reset();
+        }
+        self.condv.notify_all();
```

Comment:
if the consumed is already `bytes_per_refill`, no need to notify? or just notify_one

Reply:
Good point, this relates to a part I haven't decided yet. Say we have a thread always asking for large chunks of bytes and in the exact pace as refilling, it will eventually starve other threads, calling `notify_one` won't help either. The best approach would be setting some limits for the bytes we are giving out to each request.

Reply:
the request would be placed at the tail of the queue, so that seems it won't starve other threads?

Reply:
Not if this thread with "perfect" timing can always do the refill itself before others waking up. Also besides this case, generally it's negative for fair scheduling. e.g. queries on wide table take more quotas than those on normal table.